cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -154,115 +142,121 @@
   private static volatile boolean upgradeCompleted;
   private static long lastUpgradeCompletedCheck;
 
-  private final CConfiguration cConf;
+  private final StructuredTable applicationSpecificationTable;
+  private final StructuredTable workflowNodeStateTable;
+  private final StructuredTable runRecordsTable;
+  private final StructuredTable workflowsTable;
+  private final StructuredTable programCountsTable;
+  private final StructuredTable upgradeMetadataTable;
+  private final StructuredTable subscriberStateTable;
 
   /**
    * Static method for creating an instance of {@link AppMetadataStore}.
    */
-  public static AppMetadataStore create(CConfiguration cConf,
-                                        DatasetContext datasetContext,
-                                        DatasetFramework datasetFramework) {
+  public static AppMetadataStore create(StructuredTableContext context) {
     try {
-      Table table = DatasetsUtil.getOrCreateDataset(datasetContext, datasetFramework, APP_META_INSTANCE_ID,
-                                                    Table.class.getName(), DatasetProperties.EMPTY);
-      return new AppMetadataStore(table, cConf);
-    } catch (DatasetManagementException | IOException e) {
+      return new AppMetadataStore(context);
+    } catch (TableNotFoundException e) {
       throw new RuntimeException(e);
     }
   }
 
-  public AppMetadataStore(Table table, CConfiguration cConf) {
-    super(table);
-    this.cConf = cConf;
-  }
-
-  @Override
-  protected <T> byte[] serialize(T value) {
-    return Bytes.toBytes(GSON.toJson(value));
-  }
-
-  @Override
-  protected <T> T deserialize(MDSKey key, byte[] serialized, Type typeOfT) {
-    if (RunRecordMeta.class.equals(typeOfT)) {
-      RunRecordMeta meta = GSON.fromJson(Bytes.toString(serialized), RunRecordMeta.class);
-      meta = RunRecordMeta.builder(meta)
-        .setProgramRunId(getProgramID(key).run(meta.getPid()))
-        .build();
-      //noinspection unchecked
-      return (T) meta;
-    } else {
-      return GSON.fromJson(Bytes.toString(serialized), typeOfT);
-    }
+  public AppMetadataStore(StructuredTableContext context) throws TableNotFoundException {
+    this.applicationSpecificationTable =
```

Comment:
It seems to me that this is called in many places for every transaction. Most of the time, only one of these tables is needed. I think it would be quite expensive to always get all tables, especially if that requires looking up in the table registry. Can you change this so that each table is only instantiated when one the methods requires it?

Reply:
Do you think it would be good to define classes ApplicationTable, WorkflowNodeTable, RunRecordsTable, etc. that implement only the reading/writing/serialization of these tables. That would turn this huge class into smaller pieces

Reply:
Is it that expensive to get a table from the registry? I agree that we should turn this one class into multiple - but would rather not do so in this PR unless we must for performance reasons

Reply:
After discussion offline, we decided to dynamically load the tables when needed